### PR TITLE
[PostgreNoSQL] Added Binding to benchmark NoSQL support.

### PR DESCRIPTION
**Pull request by leschekhomann**
Added PostgreNoSQL support to YCSB. The implementation allows to insert, update, scan and read JSON documents in PostgreSQL.

### CODE REVIEW

File: postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+      } catch (Exception e) {
+        LOG.error("Error during initialization: " + e);
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    if (INIT_COUNT.decrementAndGet() == 0) {
+      postgrenosqlDriver = null;
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java]: need to call `close` on the Connection instance as a part of cleanup. if autocommit wasn't configured, you need to expressly call `commit` prior to `close` as well.

File: postgrenosql/pom.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+permissions and limitations under the License. See accompanying
+LICENSE file.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>com.yahoo.ycsb</groupId>
+    <artifactId>binding-parent</artifactId>
+    <version>0.15.0-SNAPSHOT</version>
```
Comment (busbey) [inline postgrenosql/pom.xml]: sorry! the master branch has updated to version 0.16.0-SNAPSHOT now.

File: postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+  @Override
+  public Status update(String tableName, String key, Map<String, ByteIterator> values) {
+    try{
+      JSONObject jsonObject = new JSONObject();
+      for (Map.Entry<String, ByteIterator> entry : values.entrySet()) {
+        jsonObject.put(entry.getKey(), entry.getValue().toString());
+      }
+
+      PreparedStatement updateStatement = connection.prepareStatement(createUpdateStatement(tableName));
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java]: Shouldn't these prepared statements be cached? Plenty of other implementations have left this kind of optimization for later, I'm just curious if there's some special consideration in building them each time for now.

File: postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+      return Status.UNEXPECTED_STATE;
+    } catch (SQLException e) {
+      LOG.error("Error in processing delete to table: " + tableName + e);
+      return Status.ERROR;
+    }
+  }
+
+  private String createReadStatement(String table, Set<String> fields) {
+    // TODO: Change implementation to prepared statements
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java]: is this TODO for reusing the prepared statements?

File: postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/package-info.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+/**
+ * The YCSB binding for PostgreNoSQL client.
+ */
+package com.yahoo.ycsb.postgrenosql;
```
Comment (busbey) [inline postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/package-info.java]: why the second package-info?

File: postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClientTest.java [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    props.setProperty(PostgreNoSQLDBClient.CONNECTION_URL, TEST_DB_URL);
+    props.setProperty(PostgreNoSQLDBClient.CONNECTION_USER, DEFAULT_USER);
+    props.setProperty(PostgreNoSQLDBClient.CONNECTION_PASSWD, DEFAULT_PWD);
+    props.setProperty("user", DEFAULT_USER);
+    props.setProperty("password", DEFAULT_PWD);
+    props.setProperty(PostgreNoSQLDBClient.JDBC_AUTO_COMMIT, "true");
+
+    try{
+      postgreSQLConnection = new Driver().connect(TEST_DB_URL, props);
+      deleteTable();
```
Comment (busbey) [inline postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClientTest.java]: why have these extra create / delete table calls along with the create table call in the travis script? you should choose either doing the set up in travis (and add to the class javadoc that the test presumes you've set up a correct table) or do it here in the test. Personally, I think the "presume you have a table and use travis" is easier to maintain, but I don't feel strongly.

**Comment by busbey** [comment]
Still working on this @leschekhomann?

**Comment by leschekhomann** [comment]
@busbey: Yes, I will get this done in the next two weeks. I am a little busy at the moment.

**Comment by busbey** [comment]
Great to hear! I know things get busy, so please don't feel rushed. Pings are just to manage finding when folks don't plan to follow up.

**Pull request closed without merging**